node/src/components/consensus/highway_core/state.rs

```diff
@@ -361,54 +363,47 @@ impl<C: Context> State<C> {
     pub(crate) fn validate_vote(&self, swvote: &SignedWireVote<C>) -> Result<(), VoteError> {
         let wvote = &swvote.wire_vote;
         let creator = wvote.creator;
-        wvote.panorama.validate(self)?;
-        let mut justifications = wvote.panorama.iter_correct();
-        if !justifications.all(|vh| self.vote(vh).timestamp <= wvote.timestamp) {
+        let panorama = &wvote.panorama;
+        let timestamp = wvote.timestamp;
+        panorama.validate(self)?;
+        if panorama.iter_correct(self).any(|v| v.timestamp > timestamp) {
             return Err(VoteError::Timestamps);
         }
-        match wvote.panorama.get(creator) {
-            Observation::Faulty => {
-                error!("Vote from faulty validator should be rejected in `pre_validate_vote`.");
-                return Err(VoteError::FaultyCreator); // Should be unreachable.
+        if wvote.seq_number != panorama.next_seq_num(self, creator) {
+            return Err(VoteError::SequenceNumber);
+        }
+        let opt_prev_vote = panorama[creator].correct().map(|vh| self.vote(vh));
+        let prev_round_exp = opt_prev_vote.map_or(self.params.init_round_exp(), |v| v.round_exp);
+        let max_re = prev_round_exp.max(wvote.round_exp);
+        // The round exponent must only change one step at a time.
+        if prev_round_exp + 1 < max_re || wvote.round_exp + 1 < max_re {
```

Comment:
Maybe it's just me but I find this condition difficult to follow. It might be the case that `max_re` _is_ `prev_round_exp` in which case why are we comparing it to itself? it will never return `VoteError::RoundLenght`. Actually, `prev_round_exp` cannot ever by incorrect b/c it's in the past. We should be checking only `wvote.round_exp`.

Reply:
This is the "steps of at most 1" condition, so I wanted to check that:
```rust
prev_round_exp <= wvote.round_exp + 1 && wvote.round_exp + 1 <= prev_round_exp
```
Of course `prev_round_exp` didn't violate the rule. What we require here is that `wvote.round_exp` does not differ from it by more than `1`.

Using `max_re` for this is equivalent and shorter, and we needed to compute `max_re` anyway for the later condition that round exponents only change at round boundaries. But if you prefer the above I'm happy to change it.